### PR TITLE
refactor: Cleanup Concretize module

### PR DESCRIPTION
**Pull request by scolsen**
This PR performs a few refactors in efforts to "cleanup" the `Concretize` module, whereby I mean it:

- Removes a bunch of local definitions to make functions easier to modify (no chance of accidentally altering function-local internal state)
- Adds an explicit export list to the module
- Tidies up some monadic usages to make them more terse and remove unneeded function calls.

I've also added some more pattern synonyms.

This is purely a refactor. This shouldn't result in any behavioral changes.

### CODE REVIEW

**Comment by hellerve** [comment]
I haven’t checked the PR yet. When does the new error (about type instantiation) kick in?

**Comment by scolsen** [comment]
> I haven’t checked the PR yet. When does the new error (about type instantiation) kick in?

It's in `instantiateGenericStructType` -- we use to just throw `error "instantiategenericstructtype"`

**Comment by eriksvedang** [comment]
Oh, awesome. I only did a quick read-through but it looks very promising! Will do a proper review as soon as possible.

**Review by hellerve — COMMENTED** [review]
The diff is a little hard to read, but this looks like a great refactor! Thank you so much for making the codebase more habitable 🎉

File: src/Concretize.hs [hunk 1]
```diff
@@ -36,347 +59,398 @@ data Level = Toplevel | Inside
...
+-- These functions take a state and supporting information and gradually
+-- convert generically typed xobjs into concretely typed xobjs.
+--
+-- The functions prefixed "visit" primarily just recur into the component parts
+-- of different Carp forms while the "concretizeType..." functions perform the
+-- actual type conversion work.
+
+-- | The type of visit functions. These functions convert the types of the
+-- components of a form into concrete types and take the following arguments:
+--   - A bool indicating whether or not type variables are allowed
```
Comment (eriksvedang) [inline src/Concretize.hs:78]: The first argument ([SymPath]) isn't covered in this description.

File: src/Concretize.hs [hunk 2]
```diff
@@ -36,347 +59,398 @@ data Level = Toplevel | Inside
...
-                          XObj (Arr structMemberPairs) Nothing Nothing
-                        ]
+    go b = pure [defn, name, args, b]
+visitDefn _ _ Inside _ _ x@(ListPat (DefnPat _ _ _ _)) =
+  pure (Left (DefinitionsMustBeAtToplevel x))
+visitDefn _ _ _ _ _ x = pure (Left (CannotConcretize x))
+
+-- | Concretely type a program entry point. Can only return Int or Unit.
+visitMain :: Visitor
+visitMain vs _ Toplevel tenv env (ListPat (DefnPat defn name@(SymPat (SymPath [] "main") _) args@(ArrPat []) body)) =
```
Comment (eriksvedang) [inline src/Concretize.hs:152]: The `vs` variable name is very cryptic to me, what does it stand for?

**Review by eriksvedang — COMMENTED** [review]
I didn't have time to completely go through all code right now, but I had some small initial comments.

File: src/Concretize.hs [hunk 3]
```diff
@@ -36,347 +59,398 @@ data Level = Toplevel | Inside
...
+  do
+    c <- concretizeTypeOfXObj tenv body
+    v <- visit vs False Inside tenv env body
```
Comment (eriksvedang) [inline src/Concretize.hs:155]: It really looks like `v` is related to `vs` here. I think more descriptive names for these visited forms would be very good for readability.

File: src/Concretize.hs (hunk 1 above)
Comment (scolsen) [reply src/Concretize.hs:78]: good catch! fixed

File: src/Concretize.hs (hunk 2 above)
Comment (scolsen) [reply src/Concretize.hs:152]: it's just the visitedDefinitions -- it's not used in most cases but needs to be threaded through for the one case in which it matters (mkLambda)

File: src/Concretize.hs (hunk 3 above)
Comment (scolsen) [reply src/Concretize.hs]: Good call, updated

**Pull request merged**